File: shodanify/scanner.py
"""Active, lightweight health checks for targets in the dataset.

For each target this performs a single TCP connect (reachability + latency),
one HTTP(S) GET (status, redirects, server, page title, timing) and a reverse
DNS lookup. It is deliberately minimal — one request per target — and
identifies itself with a clear User-Agent. Stdlib only.
"""
from concurrent.futures import ThreadPoolExecutor
from datetime import datetime, timezone
import re
import socket
import ssl
import time
import urllib.error
import urllib.request

USER_AGENT = "Shodanify-Healthcheck/1.0 (+defensive security outreach)"
HTTPS_PORTS = {443, 8443, 9443, 4443, 10443}
_TITLE_RE = re.compile(r"<title[^>]*>(.*?)</title>", re.I | re.S)


def _now():
    return datetime.now(timezone.utc).isoformat(timespec="seconds").replace("+00:00", "Z")


def _scheme(port):
    return "https" if port in HTTPS_PORTS else "http"


def _build_url(host, port, scheme):
    default = (scheme == "https" and port == 443) or (scheme == "http" and port == 80)
    return f"{scheme}://{host}" + ("" if default else f":{port}")


def _tcp_check(ip, port, timeout):
    start = time.monotonic()
    try:
        with socket.create_connection((ip, port), timeout=timeout):
            return True, round((time.monotonic() - start) * 1000)
    except OSError:
        return False, None


def _reverse_dns(ip):
    try:
        return socket.gethostbyaddr(ip)[0]
    except (OSError, socket.herror):
        return None


def _http_check(host, port, timeout):
    scheme = _scheme(port)
    url = _build_url(host, port, scheme)
    ctx = ssl.create_default_context()
    ctx.check_hostname = False
    ctx.verify_mode = ssl.CERT_NONE  # we want a status even from self-signed certs
    req = urllib.request.Request(url, headers={"User-Agent": USER_AGENT}, method="GET")
    start = time.monotonic()
    try:
        with urllib.request.urlopen(req, timeout=timeout, context=ctx) as resp:
            body = resp.read(20000).decode("utf-8", "ignore")
            ms = round((time.monotonic() - start) * 1000)
            final = resp.geturl()
            m = _TITLE_RE.search(body)
            return {
                "http_status": resp.status,
                "http_ok": 200 <= resp.status < 400,
                "server": resp.headers.get("Server"),
                "title": (m.group(1).strip()[:200] if m else None),
                "final_url": final,
                "redirected": final.rstrip("/") != url.rstrip("/"),
                "http_ms": ms,
                "scheme": scheme,
                "url": url,
            }
    except urllib.error.HTTPError as e:
        # A 4xx/5xx is still a live web server — record the code.
        return {
            "http_status": e.code,
            "http_ok": False,
            "server": (e.headers.get("Server") if e.headers else None),
            "title": None,
            "final_url": url,
            "redirected": False,
            "http_ms": round((time.monotonic() - start) * 1000),
            "scheme": scheme,
            "url": url,
        }
    except (urllib.error.URLError, ssl.SSLError, socket.timeout, OSError) as e:
        return {
            "http_status": None,
            "http_ok": False,
            "server": None,
            "title": None,
            "final_url": None,
            "redirected": False,
            "http_ms": None,
            "scheme": scheme,
            "url": url,
            "http_error": str(getattr(e, "reason", e))[:160],
        }


def check_target(ip, port, hostname=None, connect_timeout=4, read_timeout=6):
    """Run TCP + HTTP + rDNS checks for one target. Never raises."""
    host_for_http = hostname or ip
    tcp_open, tcp_ms = _tcp_check(ip, port, connect_timeout)
    result = {
        "ip": ip,
        "port": port,
        "hostname": hostname,
        "rdns": _reverse_dns(ip),
        "tcp_open": tcp_open,
        "tcp_ms": tcp_ms,
        "scanned_at": _now(),
    }
    if tcp_open:
        result.update(_http_check(host_for_http, port, read_timeout))
    else:
        result.update({"http_status": None, "http_ok": False, "scheme": _scheme(port),
                       "url": _build_url(host_for_http, port, _scheme(port))})
    return result


def scan_targets(targets, workers=16, connect_timeout=4, read_timeout=6):
    """Scan many targets concurrently. ``targets`` is a list of
    ``(ip, port, hostname)`` tuples. Returns a list of result dicts."""
    targets = list(targets)
    if not targets:
        return []

    def _run(t):
        ip, port, hostname = t
        return check_target(ip, port, hostname, connect_timeout, read_timeout)

    with ThreadPoolExecutor(max_workers=min(workers, len(targets))) as pool:
        return list(pool.map(_run, targets))

File: src/scanner.rs
use chrono::Utc;
use regex::Regex;
use serde_json::{json, Value};
use std::net::SocketAddr;
use std::sync::OnceLock;
use std::time::{Duration, Instant};
use tokio::net::TcpStream;
use tokio::time::timeout;

const USER_AGENT: &str = "Shodanify-Healthcheck/1.0 (+defensive security outreach)";

// Ports that are unambiguously one scheme.
const HTTP_ONLY:  &[u16] = &[80, 8080, 3000, 5000, 8000, 8888];
const HTTPS_ONLY: &[u16] = &[443, 8443, 9443, 4443, 10443];

static TITLE_RE: OnceLock<Regex> = OnceLock::new();

fn title_re() -> &'static Regex {
    TITLE_RE.get_or_init(|| Regex::new(r"(?is)<title[^>]*>(.*?)</title>").unwrap())
}

fn now_iso() -> String {
    Utc::now().format("%Y-%m-%dT%H:%M:%SZ").to_string()
}

fn build_url(host: &str, port: u16, scheme: &str) -> String {
    let is_default = (scheme == "https" && port == 443) || (scheme == "http" && port == 80);
    if is_default {
        format!("{}://{}", scheme, host)
    } else {
        format!("{}://{}:{}", scheme, host, port)
    }
}

async fn tcp_check(ip: &str, port: u16, connect_timeout: Duration) -> (bool, Option<u64>) {
    let addr: SocketAddr = match format!("{}:{}", ip, port).parse() {
        Ok(a) => a,
        Err(_) => return (false, None),
    };
    let start = Instant::now();
    match timeout(connect_timeout, TcpStream::connect(addr)).await {
        Ok(Ok(_)) => (true, Some(start.elapsed().as_millis() as u64)),
        _ => (false, None),
    }
}

async fn reverse_dns(ip: String) -> Option<String> {
    tokio::task::spawn_blocking(move || {
        dns_lookup::lookup_addr(&ip.parse().ok()?).ok()
    })
    .await
    .ok()
    .flatten()
}

// Returns true if the error is a transport/protocol mismatch (e.g. sent HTTPS to an
// HTTP server) rather than a network or application-level error.
fn is_scheme_mismatch(err: &reqwest::Error) -> bool {
    // No status code means it failed before getting an HTTP response.
    if err.status().is_some() { return false; }
    let s = err.to_string().to_lowercase();
    // TLS handshake failure when server speaks plain HTTP, or an unexpected EOF/reset
    // when a plain-HTTP client connects to a TLS port.
    s.contains("tls") || s.contains("handshake") || s.contains("corrupt")
        || s.contains("eof") || s.contains("connection reset")
        || s.contains("record overflow") || s.contains("illegal parameter")
}

// Make a single HTTP request with the given scheme; returns the result JSON.
async fn try_scheme(client: &reqwest::Client, host: &str, port: u16, scheme: &str) -> Value {
    let url = build_url(host, port, scheme);
    let start = Instant::now();

    match client.get(url.as_str()).send().await {
        Ok(resp) => {
            let status = resp.status().as_u16() as i64;
            let http_ok = (200..400).contains(&(status as u16));
            let server = resp.headers()
                .get("server")
                .and_then(|v| v.to_str().ok())
                .map(|s| s.to_string());
            let final_url = resp.url().to_string();
            let redirected = final_url.trim_end_matches('/') != url.trim_end_matches('/');

            let body = match resp.bytes().await {
                Ok(b) => {
                    let slice = if b.len() > 20_000 { &b[..20_000] } else { &b[..] };
                    String::from_utf8_lossy(slice).to_string()
                }
                Err(_) => String::new(),
            };

            let ms = start.elapsed().as_millis() as u64;
            let title = title_re()
                .captures(&body)
                .and_then(|c| c.get(1))
                .map(|m| {
                    let t = m.as_str().trim();
                    let t = if t.len() > 200 {
                        let mut idx = 200;
                        while idx > 0 && !t.is_char_boundary(idx) { idx -= 1; }
                        &t[..idx]
                    } else { t };
                    t.to_string()
                });

            json!({
                "http_status": status,
                "http_ok": http_ok,
                "server": server,
                "title": title,
                "final_url": final_url,
                "redirected": redirected,
                "http_ms": ms,
                "scheme": scheme,
                "url": url,
            })
        }
        Err(e) => {
            let ms = start.elapsed().as_millis() as u64;
            if let Some(status) = e.status() {
                json!({
                    "http_status": status.as_u16() as i64,
                    "http_ok": false,
                    "server": null, "title": null,
                    "final_url": url, "redirected": false,
                    "http_ms": ms, "scheme": scheme, "url": url,
                })
            } else {
                let err_str = e.to_string();
                let short_err = if err_str.len() > 160 { err_str[..160].to_string() } else { err_str };
                json!({
                    "http_status": null, "http_ok": false, "server": null, "title": null,
                    "final_url": null, "redirected": false,
                    "http_ms": ms, "scheme": scheme, "url": url,
                    "http_error": short_err,
                    "_scheme_mismatch": is_scheme_mismatch(&e),
                })
            }
        }
    }
}

// Tries the canonical scheme for the port; if that fails with a protocol mismatch
// (e.g. HTTPS port that happens to serve plain HTTP) it retries with the other scheme.
// For unambiguous ports (80, 443, etc.) it never retries.
async fn http_check(client: &reqwest::Client, host: &str, port: u16) -> Value {
    let primary = if HTTP_ONLY.contains(&port) {
        "http"
    } else if HTTPS_ONLY.contains(&port) {
        "https"
    } else {
        // Ambiguous port: default to https (more common for non-standard ports in Shodan data),
        // fall back to http if we get a TLS-level error.
        "https"
    };

    let result = try_scheme(client, host, port, primary).await;

    // Only retry for genuinely ambiguous ports — avoid double-requesting known ports.
    let is_ambiguous = !HTTP_ONLY.contains(&port) && !HTTPS_ONLY.contains(&port);
    if is_ambiguous {
        let mismatch = result
            .get("_scheme_mismatch")
            .and_then(|v| v.as_bool())
            .unwrap_or(false);
        if mismatch {
            let fallback = if primary == "https" { "http" } else { "https" };
            let alt = try_scheme(client, host, port, fallback).await;
            // Use fallback only if it actually got an HTTP-level response.
            if alt.get("http_status").map(|v| !v.is_null()).unwrap_or(false) {
                return alt;
            }
        }
    }

    // Strip the internal helper field before returning.
    if let Some(obj) = result.as_object() {
        let mut clean: serde_json::Map<String, Value> = obj.clone();
        clean.remove("_scheme_mismatch");
        return Value::Object(clean);
    }
    result
}

pub async fn check_target(
    client: &reqwest::Client,
    ip: &str,
    port: u16,
    hostname: Option<&str>,
    connect_timeout: Duration,
) -> Value {
    let host_for_http = hostname.unwrap_or(ip);

    // TCP connect and rDNS are independent — run them concurrently.
    let ((tcp_open, tcp_ms), rdns) = tokio::join!(
        tcp_check(ip, port, connect_timeout),
        reverse_dns(ip.to_string()),
    );

    let mut result = json!({
        "ip": ip,
        "port": port,
        "hostname": hostname,
        "rdns": rdns,
        "tcp_open": tcp_open,
        "tcp_ms": tcp_ms,
        "scanned_at": now_iso(),
    });

    let http_result = if tcp_open {
        http_check(client, host_for_http, port).await
    } else {
        let scheme = if HTTPS_ONLY.contains(&port) { "https" } else { "http" };
        json!({
            "http_status": null, "http_ok": false,
            "scheme": scheme,
            "url": build_url(host_for_http, port, scheme),
        })
    };

    if let (Some(obj), Some(http_obj)) = (result.as_object_mut(), http_result.as_object()) {
        for (k, v) in http_obj {
            obj.insert(k.clone(), v.clone());
        }
    }

    result
}

pub async fn scan_targets(
    targets: Vec<(String, u16, Option<String>)>,
    workers: usize,
    connect_timeout: Duration,
    read_timeout: Duration,
) -> Vec<Value> {
    use futures::stream::{FuturesUnordered, StreamExt};

    let client = std::sync::Arc::new(
        reqwest::ClientBuilder::new()
            .danger_accept_invalid_certs(true)
            .timeout(read_timeout)
            .redirect(reqwest::redirect::Policy::limited(10))
            .user_agent(USER_AGENT)
            .build()
            .expect("failed to build HTTP client"),
    );

    let semaphore = std::sync::Arc::new(tokio::sync::Semaphore::new(workers));
    let mut futs = FuturesUnordered::new();

    for (ip, port, hostname) in targets {
        let sem = semaphore.clone();
        let c = client.clone();
        futs.push(tokio::spawn(async move {
            let _permit = sem.acquire().await.unwrap();
            check_target(&c, &ip, port, hostname.as_deref(), connect_timeout).await
        }));
    }

    let mut results = Vec::with_capacity(futs.len());
    while let Some(res) = futs.next().await {
        if let Ok(v) = res { results.push(v); }
    }
    results
}
